x/application/keeper/transfer_applications.go

```diff
@@ -0,0 +1,241 @@
+		}
+	}
+}
+
+// MergeAppPendingUndelegations takes the union of the srcApp and dstApp's pending undelegations
+// and sets the result in dstApp. Pending undelegations are merged according to the following algorithm:
+// - At each pending undelegation height in the destination application:
+//   - Take the union of the gateway addresses in source and destination applications'
+//     pending undelegations at that height.
+//   - If a gateway address is present in both source and destination application's pending undelegations
+//     **at different heights**, the destination application's undelegation height is used
+//     (i.e. this undelegation is unchanged) and that gateway address is excluded from the
+//     gateway address union at the height which it is present in the source application's
+//     pending undelegations.
+//
+// It is exported for testing purposes.
+func MergeAppPendingUndelegations(srcApp, dstApp *types.Application) {
+	// Build a map from all gateway addresses which have pending undelegations to
+	// their respective undelegation session end heights. If the source and destination
+	// applications both contain pending undelegations from the same gateway address, the
+	// source pending undelegation is ignored. E.g., srcApp has a pending undelegation at
+	// height 10 from gateway1 and dstApp has pending undelegation from gateway1 atheight
+	// 20; only one pending undelegation from gateway1 should be present in the merged
+	// result, at height 20.
+	undelegationsUnionAddrToHeightMap := make(map[string]uint64)
+
+	for srcHeight, srcUndelegetingGatewaysList := range srcApp.PendingUndelegations {
+		for _, srcGateway := range srcUndelegetingGatewaysList.GatewayAddresses {
+			undelegationsUnionAddrToHeightMap[srcGateway] = srcHeight
+		}
+	}
+
+	for dstHeight, dstUndelegatingGatewaysList := range dstApp.PendingUndelegations {
+		for _, dstGateway := range dstUndelegatingGatewaysList.GatewayAddresses {
+			undelegationsUnionAddrToHeightMap[dstGateway] = dstHeight
+		}
+	}
+
+	// Reset the destination application's pending undelegations and rebuild it
+	// from the undelegations union map.
+	dstApp.PendingUndelegations = make(map[uint64]types.UndelegatingGatewayList)
+	for gatewayAddr, height := range undelegationsUnionAddrToHeightMap {
+		dstPendingUndelegationsAtHeight, ok := dstApp.PendingUndelegations[height]
+		if !ok {
+			dstApp.PendingUndelegations[height] = types.UndelegatingGatewayList{
+				GatewayAddresses: []string{gatewayAddr},
+			}
+		} else {
+			dstPendingUndelegationsAtHeight.GatewayAddresses = append(dstApp.PendingUndelegations[height].GatewayAddresses, gatewayAddr)
+			dstApp.PendingUndelegations[height] = dstPendingUndelegationsAtHeight
+		}
+	}
+}
+
+// MergeAppServiceConfigs takes the union of the srcApp and dstApp's service configs
+// and sets the result in dstApp. It is exported for testing purposes.
+func MergeAppServiceConfigs(srcApp, dstApp *types.Application) {
```

Comment:
Do you have a followup PR that updates the docs on this?

Would love a **simple** excalidraw/mermaid diagram for it.

Reply:
I will include it in #790. :+1: